### PR TITLE
Export execution details

### LINKED ISSUE #1137: Add messages to RaoResult
### Describe the current behavior

There is no way to add generic messages to RaoResult. If the RAO fails, the reason of the failure is accessible only after casting the RaoResult to a FailedRaoResult (#1102)

### Describe the expected behavior

Add a way to put messages in RaoResult:
- when the RAO fails, put the reason in there
- when the RAO succeeds, put the executed steps in there (instead of in enum) 

### Describe the motivation

cleaner API + removing the steps enum kaes the RaoResult independent from its implementation

### Extra Information

_No response_

### PR DESCRIPTION
**Pull request by phiedw**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

Fixes #1137 

**What kind of change does this PR introduce?**

Feature

**What is the current behavior?**

The RaoResult only exports executed steps (for non failing RAOs)

**What is the new behavior (if this is a feature change)?**
We now export either the steps executed or a failure reason (as a string). Also surronded the main parts of the RAO to be able to return a FailedRaoResult instead of crashing when an unexpected RuntimeException happens.

**Does this PR introduce a breaking change or deprecate an API?**
- [x] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**

- [x] The *Breaking Change* or *Deprecated* label has been added
- [x] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**

If you use RaoResult.getOptimizationSteps, you will have to replace it with RaoResult.getExecutionDetails

### CODE REVIEW

**Comment by phiedw** [comment]
Part of this PR can actually make debugging harder because you lose access to the stack trace. Should I add a business error log in the catch to log the error and its stack trace?

**Comment by phiedw** [comment]
> Part of this PR can actually make debugging harder because you lose access to the stack trace. Should I add a business error log in the catch to log the error and its stack trace?

Added a business error

**Review by Godelaine — CHANGES_REQUESTED** [review]

File: ra-optimisation/search-tree-rao/src/test/java/com/powsybl/openrao/searchtreerao/result/impl/OneStateOnlyRaoResultImplTest.java [hunk 1]
```diff
@@ -425,15 +425,10 @@ void testCurativeCase2() {
...
-        assertFalse(output.getOptimizationStepsExecuted().hasRunSecondPreventive());
-        output.setOptimizationStepsExecuted(OptimizationStepsExecuted.SECOND_PREVENTIVE_IMPROVED_FIRST);
-        assertTrue(output.getOptimizationStepsExecuted().hasRunSecondPreventive());
-        OpenRaoException exception = assertThrows(OpenRaoException.class, () -> output.setOptimizationStepsExecuted(OptimizationStepsExecuted.FIRST_PREVENTIVE_ONLY));
-        assertEquals("The RaoResult object should not be modified outside of its usual routine", exception.getMessage());
-        exception = assertThrows(OpenRaoException.class, () -> output.setOptimizationStepsExecuted(OptimizationStepsExecuted.SECOND_PREVENTIVE_FELLBACK_TO_FIRST_PREVENTIVE_SITUATION));
-        assertEquals("The RaoResult object should not be modified outside of its usual routine", exception.getMessage());
-        exception = assertThrows(OpenRaoException.class, () -> output.setOptimizationStepsExecuted(OptimizationStepsExecuted.FIRST_PREVENTIVE_FELLBACK_TO_INITIAL_SITUATION));
-        assertEquals("The RaoResult object should not be modified outside of its usual routine", exception.getMessage());
+        setUp();
```
[the remark this answers is not in the document]
Comment (phiedw) [reply ra-optimisation/search-tree-rao/src/test/java/com/powsybl/openrao/searchtreerao/result/impl/OneStateOnlyRaoResultImplTest.java:429]: indeed!

File: data/rao-result/rao-result-json/src/main/java/com/powsybl/openrao/data/raoresultjson/serializers/RaoResultSerializer.java [hunk 2]
```diff
@@ -45,10 +44,10 @@ public void serialize(RaoResult raoResult, JsonGenerator jsonGenerator, Serializ
         // computation status
         ComputationStatus computationStatus = raoResult.getComputationStatus();
         jsonGenerator.writeStringField(COMPUTATION_STATUS, serializeStatus(computationStatus));
+        jsonGenerator.writeStringField(EXECUTION_DETAILS, raoResult.getExecutionDetails());
 
         if (!(raoResult instanceof FailedRaoResultImpl)) {
-            OptimizationStepsExecuted optimizationStepsExecuted = raoResult.getOptimizationStepsExecuted();
-            jsonGenerator.writeStringField(OPTIMIZATION_STEPS_EXECUTED, serializeOptimizedStepsExecuted(optimizationStepsExecuted));
+            jsonGenerator.writeStringField(EXECUTION_DETAILS, raoResult.getExecutionDetails());
```
[the remark this answers is not in the document]
Comment (phiedw) [reply data/rao-result/rao-result-json/src/main/java/com/powsybl/openrao/data/raoresultjson/serializers/RaoResultSerializer.java]: I'm surprised this was not caught by a unit test, I'll add one

**Review by Godelaine — APPROVED** [review]

**Pull request merged**